uDiscovery Redesign (#219)
eclipse-uprotocol/up-spec: up-l3/udiscovery/v3/service.adoc

----
SPDX-FileCopyrightText: 2024 Contributors to the Eclipse Foundation

See the NOTICE file(s) distributed with this work for additional
information regarding copyright ownership.

This program and the accompanying materials are made available under
the terms of the Apache License Version 2.0 which is available at
https://www.apache.org/licenses/LICENSE-2.0
 
SPDX-FileType: DOCUMENTATION
SPDX-License-Identifier: Apache-2.0
----

== Overview

UDiscovery provides the means for uEs to discover services (and their metadata), for local and remote services. Implementation of uDiscovery takes inspiration from Domain Name System (DNS) principles in that we store and cache queries and replicate local information to more centralized databases. The local uDiscovery service acts as the first point of contact for discovery and then the discovery service looks up at the domain level service for information about services across a domain (i.e. vehicle). The central instance of uDiscovery service  is the last point of contact for global searches of services.

UDiscovery Service implements two sets of APIs:

> - *Client APIs (link:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto]):* Used by uEs to find services and their topic metadata, and for recursive lookups at teh domain and central levels.

Reviewer: spelling [the]
```suggestion
- *Client APIs (link:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto]):* Used by uEs to find services and their topic metadata, and for recursive lookups at teh domain and central levels.
```

> - *Client APIs (link:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto]):* Used by uEs to find services and their topic metadata, and for recursive lookups at teh domain and central levels.

Reviewer: spelling
... and for recursive lookups at [the] domain and central levels
```suggestion
- *Client APIs (link:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto]):* Used by uEs to find services and their topic metadata, and for recursive lookups at teh domain and central levels.
```

- *Replication API (link:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery_replicator.proto[udiscovery_replicator.proto]):* Used Internally by uDiscovery services to add, update, or delete service metadata from local, domain, and/or central service instances.

<<udiscovery-design>> below provides an overview of the design and use of the Client and Replication APIs to implement the uDiscovery business logic.

.uProtocol Discovery Design
[#udiscovery-design]
image::design.drawio.svg[Overview]

=== Terms

- *Fully Qualified Service URI (FQURI):* A URI that contains all the information needed to uniquely identify a service. A fully qualified service URI contains the following information:
  - *ue_id:* The unique identifier of the service (both the instance and service ID portions).
  - *authority_name:* The name of the authority that hosts the service.
  - *ue_version_major:* The version of the service.

>   - *resource_id:* The resource ID (value of 0 us used to indicate the default ID for the service).

Reviewer 2: what is the _default ID_ for a service?

Author: Resource_id of 0 means the default topic for responses and notifications to be sent to

== Client APIs

> The following section we will elaborate on the service-side requirements for the uDiscover service when it serves the client facing APIs defined in the xref:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto].

Reviewer 2: ```suggestion
In the following section we will elaborate on the service-side requirements for the uDiscover service when it serves the client facing APIs defined in the xref:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto].
```

> The following section we will elaborate on the service-side requirements for the uDiscover service when it serves the client facing APIs defined in the xref:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto].

Reviewer 2: you might want to introduce a document attribute for the link to the proto ...

=== FindServices()

> `FindServices()` is used to find the authority(ies), instance(s), and version(s) of the serice id (ue_id). The API is passed `FindServicesRequest` message and it returns a `FindServicesResponse` message that contains a batch of 1 or more services that match the search criteria.

Reviewer 2: IMHO we can omit the specific names of the request and response messages (defined in the proto anyway) and focus on their semantics.

> `FindServices()` is used to find the authority(ies), instance(s), and version(s) of the serice id (ue_id). The API is passed `FindServicesRequest` message and it returns a `FindServicesResponse` message that contains a batch of 1 or more services that match the search criteria.

Reviewer 2: ```suggestion
A client uE invokes `FindServices()` to get meta data (authority, version, instance ID) of (deployed) service instances of a given type.```

--
* *MUST* search recursively up to and including the central instance.
--

UDiscovery services perform a recursive search when the local or domain instance calls `FindService()` or `FindServiceTopics()` API and then propagates the response back to the client.

[.specitem,oft-sid="dsn~discovery-recursive-async~1",oft-needs="impl,test"]
--
* The original local search *MUST ONLY* return the response message when the recursive search has completed.
--

[.specitem,oft-sid="dsn~discovery-recursive-internal~1",oft-needs="impl,test"]
--
* The uDiscovery service *MUST* cache the results of the recursive search for the duration `ttl` that is returned in the response message.
--

.Recursive Search to Central
[#recursive-search-central]
[mermaid]
ifdef::env-github[[source,mermaid]]
----
sequenceDiagram

Client ->> Local-UDiscovery: FindService()
Local-UDiscovery ->> Domain-UDiscovery: FindService()
Domain-UDiscovery ->> Central-UDiscovery: FindService()
Central-UDiscovery -->> Domain-UDiscovery: FindServicesResponse
Domain-UDiscovery -->> Local-UDiscovery: FindServicesResponse
Local-UDiscovery -->> Client: FindServicesResponse
----

== Replication API

=== SetServiceTopics()

> The `SetServiceTopics()` API is used by the uDiscovery business logic to add, update, or remove information recursively between the local, domain, and central instances. the API is passed `SetServiceTopicsRequest` containing a repeated list of `UServiceTopicInfo` and `ttl` for when the info will expire. the function returns `SetServiceTopicsResponse` message if the method was successfully processed or it returns `UStatus` containing the error values specified below. 

Author: yes, local --> domain --> central

--
  
[.specitem,oft-sid="dsn~discovery-setservicetopic-error-permission~1",oft-needs="impl,test"]
--
* *MUST* return `UCode.PERMISSION_DENIED` if the caller is not uDiscovery service in the correct recursive order (i.e. local -> domain -> central).
* *MAY* grant access to deployment specific uEs such as a the software manager that installs or removes uEs.
--

[.specitem,oft-sid="dsn~discovery-setservicetopic-error-notfound~1",oft-needs="impl,test"]
--
* *MUST* return `UCode.NOT_FOUND` if no topic matching the URI was found when the `UServiceTopic` is being removed (i.e. `ttl=0`).
--

.Replication API Design
[#udiscovery-replication-design]
[mermaid]
ifdef::env-github[[source,mermaid]]
----
sequenceDiagram

OTAClient ->> Local-UDiscovery: SetServiceTopics()
Local-UDiscovery -->> OTAClient: SetServiceTopicsResponse

Local-UDiscovery ->> Domain-UDiscovery: SetServiceTopics()
Domain-UDiscovery -->> Local-UDiscovery: SetServiceTopicsResponse

Domain-UDiscovery ->> Central-UDiscovery: SetServiceTopics()
Central-UDiscovery -->> Domain-UDiscovery: SetServiceTopicsResponse
----

== Implementation Consideration

> It is possible the information in domain or central instances might become out of sync with what isa stored in the other instances. In order to rectify this situation:

Reviewer: spelling: ... out of sync with what [is] stored in the other instances.
```suggestion
It is possible the information in domain or central instances might become out of sync with what isa stored in the other instances. In order to rectify this situation:
```

> It is possible the information in domain or central instances might become out of sync with what isa stored in the other instances. In order to rectify this situation:

Reviewer: spelling
... might become out of sync with what [is] stored in the other instances
```suggestion
It is possible the information in domain or central instances might become out of sync with what isa stored in the other instances. In order to rectify this situation:
```

--
* Domain and central instances *MUST* ensure that the information within its database is accurate and up to date and reconcile any differences periodically using the client and replication APIs described in this specification.
--

[.specitem,oft-sid="dsn~discovery-data-reconciliation-frequency~1",oft-needs="impl,test"]
--
* *MUST* provide an implementation of data reconciliation using the above APIs where the frequency of reconciliation is customizable so the deployment can adjust the frequency based on the deployment needs.
--

[.specitem,oft-sid="dsn~discovery-data-reconciliation-data~1",oft-needs="impl,test"]
--
* When reconciling data, the data fetched *MUST* be used in lieu of current cached data. 
--

[.specitem,oft-sid="dsn~discovery-data-reconciliation-not-found~1",oft-needs="impl,test"]
--
* If `GetServiceTopics()` returns `UCode.NOT_FOUND` for data we had in our local instance, we *MUST* flush the local cache and then replicate the deletion. we fetch information to reconcile the data If the reconciliation fails, the service *MUST* log the error and continue to use the current cached data.
--

> <<udiscovery-reconciliation>> below provides an example of the domain instance reconciling with local instance and then propagating the change to the central instance. In the example the reconciliation (determining if the data is out of sync or not) happens in the `ReconcileData()` function.

Reviewer: the reconciliation (determining if the data is out of sync or not)
Will this function perform some sort of scan to inspect all of the records and check for differences?

Author: yes I assume yes.

.Reconciliation Example
[#udiscovery-reconciliation]
[mermaid]
ifdef::env-github[[source,mermaid]]
----
sequenceDiagram

participant Local-UDiscovery
participant Domain-UDiscovery
participant Central-UDiscovery

Domain-UDiscovery ->> Local-UDiscovery: GetServiceTopics()
Local-UDiscovery -->> Domain-UDiscovery: GetServiceTopicsResponse

> Domain-UDiscovery ->> Domain-UDiscovery: ReconcileData()

Author: I wonder if that is implementation details for a given implementation or something that we need to specify here. @kz9nyy Could we elaborate on the reconcileData() functionality in a subsequent PR?

=== FindServiceTopics()

> `FindServiceTopics()` is used to fetch metadata about one or more topic based on the passed `UUri`. Clients call this API to know more metadata about the topic(s) such as the format, id, permission level, etc... .

Reviewer 2: ```suggestion
`FindServiceTopics()` is used to fetch metadata about one or more topics based on the passed `UUri`. Clients call this API to get additional information about the topic(s) such as the message format, id, permission level, etc.
```
